perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -277,82 +303,111 @@ public static boolean isAuthorizedForAttribute(PerunSession sess, ActionType act
 				}
 			}
 		} else if (member != null && group != null) {
-			if (roles.contains(Role.VOADMIN)) {
+			if (roles.containsKey(Role.VOADMIN)) {
 				if (isAuthorized(sess, Role.VOADMIN, member)) return true;
 			}
-			if (roles.contains(Role.VOOBSERVER)) {
+			if (roles.containsKey(Role.VOOBSERVER)) {
 				if (isAuthorized(sess, Role.VOOBSERVER, member)) return true;
 			}
-			if (roles.contains(Role.SELF)) {
-				if (isAuthorized(sess, Role.SELF, member)) return true;
+			if (roles.containsKey(Role.SELF)) {
+				if (roles.get(Role.SELF).contains(ActionType.READ_PUBLIC) || roles.get(Role.SELF).contains(ActionType.WRITE_PUBLIC)) return true;
+				if (roles.get(Role.SELF).contains(ActionType.READ) || roles.get(Role.SELF).contains(ActionType.WRITE)) {
+					if (isAuthorized(sess, Role.SELF, member)) return true;
+				}
+				if (roles.get(Role.SELF).contains(ActionType.READ_VO) || roles.get(Role.SELF).contains(ActionType.WRITE_VO)) {
+					if (getPerunBl().getMembersManagerBl().getMemberVo(sess, member).equals(getPerunBl().getGroupsManagerBl().getVo(sess, group))) {
+						return true;
+					}
+				}
 			}
-			if (roles.contains(Role.GROUPADMIN)) {
+			if (roles.containsKey(Role.GROUPADMIN)) {
 				if (isAuthorized(sess, Role.GROUPADMIN, group)) return true;
 			}
 		} else if (user != null) {
-			if (roles.contains(Role.SELF)) if (isAuthorized(sess, Role.SELF, user)) return true;
-			if (roles.contains(Role.VOADMIN)) {
+			if (roles.containsKey(Role.SELF)) {
+				if (roles.get(Role.SELF).contains(ActionType.READ_PUBLIC) || roles.get(Role.SELF).contains(ActionType.WRITE_PUBLIC)) return true;
+				if (roles.get(Role.SELF).contains(ActionType.READ) || roles.get(Role.SELF).contains(ActionType.WRITE)) {
+					if (isAuthorized(sess, Role.SELF, member)) return true;
+				}
+				if (roles.get(Role.SELF).contains(ActionType.READ_VO) || roles.get(Role.SELF).contains(ActionType.WRITE_VO)) {
+					List<Vo> vosFromUser = getPerunBl().getUsersManagerBl().getVosWhereUserIsMember(sess, user);
+					List<Vo> vosFromCurrentUser = getPerunBl().getUsersManagerBl().getVosWhereUserIsMember(sess, sess.getPerunPrincipal().getUser());
```

Comment:
Here you also need to check if the user from Principal is not null. Otherwise, a NullPointerException might be thrown here.